[Bug]: /inference/v1/generate returns HTTP 200 on internal generation errors instead of 500

### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
DEBUG 07-20 09:32:52 [plugins/__init__.py:36] No plugins for group vllm.platform_plugins found.
DEBUG 07-20 09:32:52 [platforms/__init__.py:36] Checking if TPU platform is available.
DEBUG 07-20 09:32:52 [platforms/__init__.py:55] TPU platform is not available because: No module named 'libtpu'
DEBUG 07-20 09:32:52 [platforms/__init__.py:61] Checking if CUDA platform is available.
DEBUG 07-20 09:32:52 [platforms/__init__.py:88] Exception happens when checking CUDA platform: NVML Shared Library Not Found
DEBUG 07-20 09:32:52 [platforms/__init__.py:105] CUDA platform is not available because: NVML Shared Library Not Found
DEBUG 07-20 09:32:52 [platforms/__init__.py:112] Checking if ROCm platform is available.
DEBUG 07-20 09:32:52 [platforms/__init__.py:126] ROCm platform is not available because: No module named 'amdsmi'
DEBUG 07-20 09:32:52 [platforms/__init__.py:133] Checking if XPU platform is available.
DEBUG 07-20 09:32:52 [platforms/__init__.py:164] Checking if CPU platform is available.
DEBUG 07-20 09:32:52 [platforms/__init__.py:168] Confirmed CPU platform is available because vLLM is built with CPU.
DEBUG 07-20 09:32:52 [platforms/__init__.py:164] Checking if CPU platform is available.
DEBUG 07-20 09:32:52 [platforms/__init__.py:168] Confirmed CPU platform is available because vLLM is built with CPU.
DEBUG 07-20 09:32:52 [platforms/__init__.py:245] Automatically detected platform cpu.
Collecting environment information...
==============================
        System Info
==============================
OS                           : Ubuntu 22.04.5 LTS (x86_64)
GCC version                  : (Ubuntu 12.3.0-1ubuntu1~22.04.3) 12.3.0
Clang version                : Could not collect
CMake version                : Could not collect
Libc version                 : glibc-2.35

==============================
       PyTorch Info
==============================
PyTorch version              : 2.11.0+cpu
Is debug build               : False
CUDA used to build PyTorch   : None
ROCM used to build PyTorch   : N/A
XPU used to build PyTorch    : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.13 (main, Jun 23 2026, 15:18:55) [Clang 22.1.3 ] (64-bit runtime)
Python platform              : Linux-5.14.0-687.26.1.el9_8.x86_64-x86_64-with-glibc2.35
    

==============================
          CPU Info
==============================
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           40 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  32
On-line CPU(s) list:                     0-31
Vendor ID:                               GenuineIntel
Model name:                              Intel Xeon Processor (Cascadelake)
CPU family:                              6
Model:                                   85
Thread(s) per core:                      2
Core(s) per socket:                      8
Socket(s):                               2
Stepping:                                6
BogoMIPS:                                4788.62
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat vnmi pku ospke avx512_vnni md_clear arch_capabilities
Virtualization:                          VT-x
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               1 MiB (32 instances)
L1i cache:                               1 MiB (32 instances)
L2 cache:                                64 MiB (16 instances)
L3 cache:                                32 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-15
NUMA node1 CPU(s):                       16-31
Vulnerability Gather data sampling:      Unknown: Dependent on hypervisor status
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

==============================
Versions of relevant libraries
==============================
[pip3] numpy==2.3.5
[pip3] pyzmq==27.1.0
[pip3] torch==2.11.0+cpu
[pip3] torchaudio==2.11.0+cpu
[pip3] torchcodec==0.14.0+cpu
[pip3] torchvision==0.26.0+cpu
[pip3] transformers==5.13.1
[pip3] triton==3.7.0+git270e696d
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.25.1
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled; XPU: Disabled
GPU Topology:
  Could not collect

==============================
     Environment Variables
==============================
VLLM_TARGET_DEVICE=cpu
VLLM_LOGGING_LEVEL=DEBUG
MAX_JOBS=16
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_root


```

</details>


### 🐛 Describe the bug

The scale-out /inference/v1/generate non-streaming path returns HTTP 200 with finish_reason="error" when a request fails internally during generation, instead of HTTP 500.

What was run

Started a CPU vllm-openai-cpu server for Qwen/Qwen3-VL-2B-Instruct:

```bash
docker run -d \
    --name vllm-decode \
    --network "host" \
    -e VLLM_TARGET_DEVICE=cpu \
    -e VLLM_LOGGING_LEVEL=DEBUG \
    -p 8003:8003 \
    "vllm/vllm-openai-cpu:latest" \
    --model Qwen/Qwen3-VL-2B-Instruct \
    --port 8003 \
    --block-size 16 \
    --enforce-eager \
    --max-model-len 2048 \
    --max-num-seqs 12 \
    --gpu-memory-utilization 0.15 \
    --mm-processor-kwargs '{"max_pixels": 12544}'
```

Took a kwargs=0 multimodal generate body (no features.kwargs_data), swapped in an image hash the encoder cache does not hold, and POSTed it to the non-streaming endpoint:

```bash
# take a kwargs=0 generate body (no features.kwargs_data) and swap in a hash the EC can't hold
jq '.features.mm_hashes.image=["deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef"]' \
  generate_body_mm0.json > miss.json
```
```
curl -sS http://localhost:8003/inference/v1/generate \
  -H 'Content-Type: application/json' -d @miss.json -w '\nHTTP %{http_code}\n'
```
Result
```
{"request_id":"b606fabf06a699b7","choices":[{"index":0,"finish_reason":"error","token_ids":[]}],...}
HTTP 200
```

The engine logged the failure, but the API server still returned 200 OK:
```
(EngineCore pid=436) ERROR 07-20 06:56:35 [v1/engine/core.py:1569] Unexpected error pre-processing request generate-tokens-97bc21f61aa76ba9-8f8a2096
(EngineCore pid=436) ERROR 07-20 06:56:35 [v1/engine/core.py:1569] Traceback (most recent call last):
(EngineCore pid=436) ERROR 07-20 06:56:35 [v1/engine/core.py:1569]   File ".../vllm/v1/engine/core.py", line 1481, in process_input_sockets
(EngineCore pid=436) ERROR 07-20 06:56:35 [v1/engine/core.py:1569]     request = self.preprocess_add_request(req)
(EngineCore pid=436) ERROR 07-20 06:56:35 [v1/engine/core.py:1569]   File ".../vllm/v1/engine/core.py", line 807, in preprocess_add_request
(EngineCore pid=436) ERROR 07-20 06:56:35 [v1/engine/core.py:1569]     request.mm_features = self.mm_receiver_cache.get_and_update_features(...)
(EngineCore pid=436) ERROR 07-20 06:56:35 [v1/engine/core.py:1569]   File ".../vllm/multimodal/cache.py", line 607, in get_and_update_features
(EngineCore pid=436) ERROR 07-20 06:56:35 [v1/engine/core.py:1569]     feature.data = self.get_and_update_item(feature.data, cache_key)
(EngineCore pid=436) ERROR 07-20 06:56:35 [v1/engine/core.py:1569]   File ".../vllm/multimodal/cache.py", line 660, in get_and_update_item
(EngineCore pid=436) ERROR 07-20 06:56:35 [v1/engine/core.py:1569]     assert mm_item is not None, f"Expected a cached item for {mm_hash=}"
(EngineCore pid=436) ERROR 07-20 06:56:35 [v1/engine/core.py:1569] AssertionError: Expected a cached item for mm_hash='deadbeef...deadbeef'
(APIServer pid=1) INFO:     127.0.0.1:41576 - "POST /inference/v1/generate HTTP/1.1" 200 OK
```
For contrast, the same body with valid features.kwargs_data, or the same hash once warmed, returns finish_reason":"length" at HTTP 200.

[miss.json](https://github.com/user-attachments/files/30181449/miss.json)

Expected behavior

finish_reason="error" should be surfaced as HTTP 500, matching the OpenAI-compatible endpoints (_raise_if_error).

--- a/tests/entrypoints/scale_out/token_in_token_out/test_generate_stream.py
+++ b/tests/entrypoints/scale_out/token_in_token_out/test_generate_stream.py
@@ -9,7 +9,7 @@ from unittest.mock import AsyncMock, MagicMock
 import pytest
 
 from vllm.config.multimodal import MultiModalConfig
-from vllm.entrypoints.openai.engine.protocol import StreamOptions
+from vllm.entrypoints.openai.engine.protocol import GenerationError, StreamOptions
 from vllm.entrypoints.openai.models.protocol import BaseModelPath
 from vllm.entrypoints.openai.models.serving import OpenAIServingModels
 from vllm.entrypoints.scale_out.token_in_token_out.protocol import (
@@ -227,6 +227,30 @@ async def test_serve_tokens_threads_session_id_header_to_engine():
     await serving.serve_tokens(request, raw_request)
 
     assert engine.generate.call_args.kwargs["session_id"] == "header-session"
+
+
+@pytest.mark.asyncio
+async def test_non_stream_error():
+    """finish_reason='error' raises for the HTTP exception handler."""
+    engine = _mock_engine()
+
+    async def mock_generate(*args, **kwargs):
+        yield _make_request_output(
+            "req-1", token_ids=[], finish_reason="error", finished=True
+        )
+
+    engine.generate = MagicMock(side_effect=mock_generate)
+    serving = _build_serving_tokens(engine)
+
+    request = GenerateRequest(
+        token_ids=[1, 2, 3],
+        sampling_params=SamplingParams(max_tokens=10),
+        model=MODEL_NAME,
+        stream=False,
+    )
+
+    with pytest.raises(GenerationError):
+        await serving.serve_tokens(request)
 
 
 @pytest.mark.asyncio

--- a/vllm/entrypoints/scale_out/token_in_token_out/serving.py
+++ b/vllm/entrypoints/scale_out/token_in_token_out/serving.py
@@ -266,6 +266,8 @@ class ServingTokens(GenerateBaseServing):
         choices: list[GenerateResponseChoice] = []
         num_generated_tokens = 0
         for output in final_res.outputs:
+            self._raise_if_error(output.finish_reason, request_id)
+
             token_ids = output.token_ids
             out_logprobs = output.logprobs
 

